### PR TITLE
Add create, edit and delete command functionality

### PR DESCRIPTION
**Pull request by change-my-username**

### CODE REVIEW

File: src/edit-cusom-command/edit-custom-command.action.tsx [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { Action, Icon } from "@raycast/api";
+import { CustomCommandView } from "../components/custom-command-view";
+
+interface IProps {
+  id: string;
+}
+
+export const EditCustomCommandAction = ({ id }: IProps) => {
+  return <Action.Push title="Edit Command" target={<CustomCommandView isEdit id={id} />} icon={Icon.Pencil} />;
```
Comment (cyxn) [inline src/edit-cusom-command/edit-custom-command.action.tsx]: need to add `shortcut` prop. maybe cmd + e

File: src/remove-custom-command/remove-custom-command.action.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { Action, Alert, confirmAlert, Icon, LocalStorage, popToRoot, showToast, Toast } from "@raycast/api";
+
+interface IProps {
+  id: string;
+}
+
+export const RemoveCustomCommandAction = ({ id }: IProps) => {
+  return (
```
Comment (cyxn) [inline src/remove-custom-command/remove-custom-command.action.tsx:8]: need to add `shortcut` — control + x

File: src/create-custom-command/create-custom-command.action.tsx [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { Action, Icon } from "@raycast/api";
+import { CustomCommandView } from "../components/custom-command-view";
+
+export const CreateCustomCommandAction = () => {
+  return <Action.Push title="Create Command" target={<CustomCommandView />} icon={Icon.Plus} />;
```
Comment (cyxn) [inline src/create-custom-command/create-custom-command.action.tsx]: need to add shortcut — cmd + N

File: src/manage-custom-commands.tsx [hunk 4]
```diff
@@ -127,11 +137,9 @@ export default function Command() {
                   />
```
Comment (cyxn) [inline src/manage-custom-commands.tsx:141]: let's make cmd + enter to save as quicklink and enter (primary) action to run the command

File: src/manage-custom-commands.tsx [hunk 5]
```diff
@@ -36,19 +39,26 @@ export default function Command() {
...
         preferenceBrowserName: browser?.name || "",
       });
-      let activeTabUrl = null;
+      let activeTabUrl: State["activeTabUrl"] = null;
       if (isInBrowserNow) {
         activeTabUrl = await getActiveTabUrl(frontmostApplicationName);
       }
 
-      const storedData = [applicationShortcutRaiseHandForMeet, applicationShortcutRecordToggleMute];
+      const storedData = [applicationShortcutRaiseHandForMeet, applicationShortcutRecordToggleMute, ...allCommands];
```
Comment (cyxn) [inline src/manage-custom-commands.tsx:61]: I'm thinking what should we do with pre-defined shortcuts. Maybe we should do nothing but let's add on edit screen that this is not editable and any changes will be reverted to defaults

File: src/run-custom-command.tsx [hunk 6]
```diff
@@ -1,21 +1,30 @@
...
 export default async function main({ arguments: { id } }: { arguments: { id: string } }) {
   const frontmostApplication = await getFrontmostApplication();
   const { name: frontmostApplicationName } = frontmostApplication;
   const { browser } = getPreferenceValues();
 
-  // todo: look in cache or in local storage by id;
   console.log({ id });
-  const applicationCommandRecord = idToCommandMap[id];
+
+  const allCommandsRaw = await LocalStorage.allItems();
```
Comment (cyxn) [inline src/run-custom-command.tsx]: maybe can be replaced to `getItem`? 
- First check if it is pre-defined command
- if nothing found, `getItem` should be called

File: src/components/custom-command-view/custom-command-view.tsx [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+          title={pickedApp}
+          key={pickedApp}
+          onChange={onAppShortcutChangeHandler(pickedApp)}
+          value={pickedAppShortcuts[pickedApp] || []}
+          error={errors[pickedApp]}
+        />
+      ))}
+      <Form.TextField
+        id={FormFields.urls}
+        title="Urls"
```
Comment (cyxn) [inline src/components/custom-command-view/custom-command-view.tsx]: "URLs of web applications"

File: src/components/custom-command-view/custom-command-view.tsx [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        id={FormFields.commandName}
+        title="Command Name"
+        placeholder="For example: Toggle Mute/Unmute"
+        onChange={onCommandNameChangeHandler}
+        value={commandName}
+        error={errors[FormFields.commandName]}
+      />
+      <Form.TagPicker
+        id={FormFields.apps}
+        title="Apps"
```
Comment (cyxn) [inline src/components/custom-command-view/custom-command-view.tsx]: Applications

File: src/components/custom-command-view/custom-command-view.tsx [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        placeholder="Choose apps"
```
Comment (cyxn) [inline src/components/custom-command-view/custom-command-view.tsx]: Choose applications

File: src/utils.ts [hunk 10]
```diff
@@ -86,3 +86,84 @@ export function compareIsAppMatches({
...
   return false;
 }
+
+const isModifier = (val: any): val is Modifier => modifiers.includes(val);
+
+const isKey = (val: any): val is Key => Object.keys(keyToAppleScriptCode).includes(val);
+
+const isValidShortcut = (shortcut: string[]): boolean => {
+  return shortcut.some(isModifier) && shortcut.some(isKey);
+};
```
Comment (cyxn) [inline src/utils.ts]: empty modifier should be allowed, i.e. "A" itself is valid shortcut

**Review by cyxn — CHANGES_REQUESTED** [review]

File: src/manage-custom-commands.tsx (hunk 5 above)
Comment (change-my-username) [reply src/manage-custom-commands.tsx:61]: There is no ability to mix Form and Detail view, but I add confirmation alert for this.

File: src/manage-custom-commands.tsx (hunk 4 above)
Comment (change-my-username) [reply src/manage-custom-commands.tsx:141]: 1. There is no ability to run command on enter
2. From the docs `Keep in mind that while you can specify an alternative shortcut for the primary and secondary actions, it won't be displayed in the Action Panel.` So, do we need to change it ?

File: src/utils.ts [hunk 11]
```diff
@@ -130,7 +134,8 @@ const parseShortcuts = (shortcut: string[]) => {
 
       return acc;
     },
-    { shortcutModifiers: [] } as any,
+    // there is a bug related to this. You can add several keys (non-modifiers) but only 1 will be displayed
+    { shortcutModifiers: [] as Modifier[] },
```
Comment (cyxn) [inline src/utils.ts]: self explaining

**Pull request merged**